src/utils/order_validation_utils.ts

```diff
@@ -0,0 +1,89 @@
+
+        if (isMakerTokenZRX) {
+            const requiredMakerAmount = fillMakerAmount.plus(signedOrder.makerFee);
+            if (requiredMakerAmount.greaterThan(makerBalance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerBalance);
+            }
+            if (requiredMakerAmount.greaterThan(makerAllowance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerAllowance);
+            }
+        } else {
+            const makerZRXBalance = await tokenWrapper.getBalanceAsync(zrxTokenAddress, signedOrder.maker);
+            const makerZRXAllowance = await tokenWrapper.getProxyAllowanceAsync(zrxTokenAddress, signedOrder.maker);
+
+            if (fillMakerAmount.greaterThan(makerBalance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerBalance);
+            }
+            if (fillMakerAmount.greaterThan(makerAllowance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerAllowance);
+            }
+            if (signedOrder.makerFee.greaterThan(makerZRXBalance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerFeeBalance);
+            }
+            if (signedOrder.makerFee.greaterThan(makerZRXAllowance)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerFeeAllowance);
+            }
+        }
+    }
+    private static async validateFillOrderTakerBalancesAndAllowancesAndThrowIfInvalidAsync(
+        tokenWrapper: TokenWrapper, signedOrder: SignedOrder, fillTakerAmount: BigNumber.BigNumber,
+        senderAddress: string, zrxTokenAddress: string,
+    ): Promise<void> {
+        const takerBalance = await tokenWrapper.getBalanceAsync(signedOrder.takerTokenAddress, senderAddress);
+        const takerAllowance = await tokenWrapper.getProxyAllowanceAsync(signedOrder.takerTokenAddress, senderAddress);
+
+        const isTakerTokenZRX = signedOrder.takerTokenAddress === zrxTokenAddress;
+
+        if (isTakerTokenZRX) {
+            const requiredTakerAmount = fillTakerAmount.plus(signedOrder.takerFee);
+            if (requiredTakerAmount.greaterThan(takerBalance)) {
+                throw new Error(ExchangeContractErrs.InsufficientTakerBalance);
+            }
+            if (requiredTakerAmount.greaterThan(takerAllowance)) {
+                throw new Error(ExchangeContractErrs.InsufficientTakerAllowance);
+            }
+        } else {
+            const takerZRXBalance = await tokenWrapper.getBalanceAsync(zrxTokenAddress, senderAddress);
+            const takerZRXAllowance = await tokenWrapper.getProxyAllowanceAsync(zrxTokenAddress, senderAddress);
+
+            if (fillTakerAmount.greaterThan(takerBalance)) {
+                throw new Error(ExchangeContractErrs.InsufficientTakerBalance);
+            }
+            if (fillTakerAmount.greaterThan(takerAllowance)) {
+                throw new Error(ExchangeContractErrs.InsufficientTakerAllowance);
+            }
```

Comment:
We can remove some code duplication here by setting the amount being checked against the balance/allowance to either be equal to `fillTakerAmount.plus(signedOrder.takerFee)` or `fillTakerAmount ` and then only having the conditional checks for balance/allowance once instead of twice.